Add NdBSpline: n-dim tensor product b-spline object

ev-br opened the pull request:
#### Reference issue
<!--Example: Closes gh-WXYZ.-->

cross-ref https://github.com/scipy/scipy/issues/18010, https://github.com/scipy/scipy/pull/7903

#### What does this implement/fix?
<!--Please explain your changes.-->

Add a new class `NdBSpline`, to represent an N-dimensional tensor product b-spline. This class only knows how to evaluate a tensor product given coefficients and knot vectors. This way it generalizes BSpline for 1D data to N-D, and parallels NdPPoly (which represents N-D tensor product polynomials in the power basis).

This class is intended to serve as a basis for a further class/routine to construct an N-D interpolator given multidim data, possibly internally in RegularGridInterpolator. The latter currently uses a recursive application of `make_interp_spline`, which is very slow. 
However constructing an interpolator from data is separate and is not a part of this PR. 

One other thing we get for free from this class is evaluation of gradients, cf https://github.com/scipy/scipy/pull/7903 --- which gets messy for the recursive approach.

#### Additional information
<!--Any additional information you think is important.-->

This class uses the localized nature of b-splines, and goes to some lengths to accomodate a general ND case. 
A straightforward implementation is available in https://github.com/ev-br/ndbsplne_playground/blob/main/NdBSpline_playground.ipynb , see `NdBSpline` and `NdBSpline0` classes there. 

This might be of interest to @Kai-Striega @j-bowhay and @AtsushiSakai who worked on similar things recently.
I'll also take liberty to ping @charris who discussed these things before.

Kai-Striega commented on scipy/interpolate/_bspl.pyx:
```diff
@@ -473,3 +475,193 @@ def _make_design_matrix(const double[::1] x,
+        # the number of non-zero terms for each point in ``xi``.
+        volume = 1
+        for d in range(ndim):
+            volume *= k[d] + 1
```
Can this loop be included in the `nogil` section?

ev-br replied:
Yup, moved.

Kai-Striega commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,210 @@
+            k = (k,)*ndim
+
+        if len(k) != ndim:
+            raise ValueError(f"len(t) = {ndim} != {len(k)} = len(k)")
```
FYI: this can be simplified a little bit by using `=` in the expression. See [this](https://docs.python.org/3/whatsnew/3.8.html#f-strings-support-for-self-documenting-expressions-and-debugging) entry in the Python 3.8 release notes. So now the expression could be `f"{len(t) = } != {len(k) = }"`. 

I'm not sure if it's any neater but it could be something you didn't know of.

ev-br replied:
How, I did not know about it. Thank you! Have used it in the last update.

Kai-Striega commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,210 @@
+            or an array with the shape (num_points, ndim).
+        nu : array_like, optional, shape (ndim,)
+            Orders of derivatives to evaluate. Each must be non-negative.
+        extrapolate : book, optional
```
Should this be `bool`?

Kai-Striega commented on scipy/interpolate/tests/test_bsplines.py:
```diff
@@ -1664,3 +1669,460 @@ def test_weighted_smoothing_spline(self):
+        assert_allclose(bspl2(xi),
+                        target, atol=1e-14)
+
+    def test_2D_derivative(self):
```
Could this test be parameterised?

ev-br replied:
In principle, it could. I fail to see how the result is more readable than what's here now, but won't oppose a PR :-).

Kai-Striega commented on scipy/interpolate/tests/test_bsplines.py:
```diff
@@ -1664,3 +1669,460 @@ def test_weighted_smoothing_spline(self):
+        assert result.shape == (11,)
+        assert_allclose(result, target, atol=1e-14)
+
+    def test_3D_derivative(self):
```
Could this test ve parameterised?

Kai-Striega reviewed (commented):
Looks really cool. Well done! 

I don't have the time for a full review tonight, but I'll try to get one done this weekend.

One thing: I don't have much of a background in this; do you have any resources to get an understanding of the maths going on here?

j-bowhay commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,210 @@
+        for d in range(ndim-1, -1, -1):
+            strides_c1[d] = strides_c1[d+1] * c1.shape[d+1]
+
+        assert strides_c1 == [_//c1.dtype.itemsize for _ in c1.strides]
```
Assuming these assert statements are to be removed at some point before merging?

ev-br replied:
Thanks, good point. Removed.

ev-br commented:
> One thing: I don't have much of a background in this; do you have any resources to get an understanding of the maths going on here?

Do you mean the general math behind tensor products or the tricks this PR does with jumping through the hoops of a dealing with arrays of runtime-defined dimensionality? For the general background, Chap 7 of Ref [1] of BSpline docstring is nice (https://www.uio.no/studier/emner/matnat/ifi/nedlagte-emner/INF-MAT5340/v05/undervisningsmateriale/hele.pdf) these lecture notes did help me a lot when I first looked at this; For what' s done here I just added some blurb to  the top of https://github.com/ev-br/ndbsplne_playground/blob/main/NdBSpline_playground.ipynb, let me know if it's what you're looking for and what's missing.

ev-br commented:
No idea what mypy wants from me. 
https://github.com/scipy/scipy/pull/18492/commits/4549d9eb2c81f2fe461db60cfa74e174ca6a8f03 does not seem to fix it.

Kai-Striega commented:
> Do you mean the general math behind tensor products or the tricks this PR does with jumping through the hoops of a dealing with arrays of runtime-defined dimensionality?

I meant the general maths. I haven't done any real work with splines before and would like to learn more. I'll take a look at the link and get back to you. It does seem like a good introduction.

Kai-Striega commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,204 @@
+    t : tuple of ndarrays
+        Knots vectors.
+    c : ndarray
+        Coefficients of the tensor-produce spline.
```
```suggestion
        Coefficients of the tensor-product spline.
```

I'm not too familiar with the maths, but should this be `product`?

ev-br replied:
Indeed, thanks! :-)

Kai-Striega commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,204 @@
+        nu : array_like, optional, shape (ndim,)
+            Orders of derivatives to evaluate. Each must be non-negative.
+        extrapolate : bool, optional
+            Whether to exrapolate based on first and last intervals in each
```
```suggestion
            Whether to extrapolate based on first and last intervals in each
```

Kai-Striega commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,204 @@
+        # XXX: is c1r guaranteed to be C-contiguous? Cython code assumes it is.
+
+        # replacement for np.ravel_multi_index for indexing of `c1`:
+        _strides_c1 = np.asarray([_//c1.dtype.itemsize for _ in c1.strides], dtype=np.intp)
```
This is a style issue, but it annoys me. `_` is usually used to denote a throwaway or unused variable. In this case, it's not. While it has no effect on how the code executes, I think of it's use as an antipattern.

Kai-Striega reviewed (commented):
I still haven't had the time to review the maths literature. But the code looks good to me. I would prefer it if someone familiar with the maths could give this a review before we merge it.

Kai-Striega commented:
@ev-br any progress on getting mypy happy? I think this is a mypy issue, not an issue with your code. If you need to, you can append `# type: ignore` to line 6 to make the tests pass. We should probably make sure it's documented somewhere.

Kai-Striega commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,205 @@
+    See Also
+    --------
+    BSpline : a one-dimensional B-spline object
+    NdPPoly : an N-dimensional piecewise tensor product polynomal
```
```suggestion
    NdPPoly : an N-dimensional piecewise tensor product polynomial
```

ev-br commented:
> any progress on getting mypy happy? I think this is a mypy issue, not an issue with your code. If you need to, you can append # type: ignore to line 6 to make the tests pass. We should probably make sure it's documented somewhere.

I rebased on main and the lint CI seems to pass now. Documenting this would be great actually.

Kai-Striega commented:
@ev-br is there anything else that you are wanting to add to this PR, or are we only waiting on a reviewer for the maths?

AtsushiSakai commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,205 @@
+
+    Parameters
+    ----------
+    c : ndarray, shape (n1, n2, ..., nN, ...)
```
The order of the arguments in this class's __init__ is t, c, k, so shouldn't the order of the arguments in this docstring be the same?

AtsushiSakai commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,205 @@
+        Returns
+        -------
+        values : ndarray, shape xi.shape[:-1] + self.c.shape[ndim:]
+            Interpolated values at xi
```
```suggestion
            Interpolated values at ``xi``
```

AtsushiSakai commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,205 @@
+        self.c = np.ascontiguousarray(self.c, dtype=dt)
+
+    def __call__(self, xi, nu=None, extrapolate=None):
+        """Evaluate the tensor product b-spline at `xi`.
```
```suggestion
        """Evaluate the tensor product b-spline at ``xi``.
```

AtsushiSakai commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,205 @@
+
+        Returns
+        -------
+        values : ndarray, shape xi.shape[:-1] + self.c.shape[ndim:]
```
```suggestion
        values : ndarray, shape ``xi.shape[:-1] + self.c.shape[ndim:]``
```

AtsushiSakai commented:
Sorry for my late review. My comments are only related to docs, so can be ignored to merge.

ev-br commented:
Thank you @AtsushiSakai , the last commit addresses your suggestions.

@Kai-Striega : 
> is there anything else that you are wanting to add to this PR, or are we only waiting on a reviewer for the maths?

I think this PR is done from my side. Constructing an interpolator from data + retrofitting it into RegularGridInterpolator can I think be done separately.  If there is a preference the contrary, this PR can of course wait for that other part to be implemented.

j-bowhay commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,205 @@
+        Coefficients of the tensor-produce spline.
+    k : tuple of integers
+        Degrees for each dimension.
+    extrapolate : bool
```
```suggestion
    extrapolate : bool, optional
```

j-bowhay commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,205 @@
+        dt = _get_dtype(self.c.dtype)
+        self.c = np.ascontiguousarray(self.c, dtype=dt)
+
+    def __call__(self, xi, nu=None, extrapolate=None):
```
```suggestion
    def __call__(self, xi, *, nu=None, extrapolate=None):
```

j-bowhay commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,205 @@
+            This can be a list or tuple of ndim-dimensional points
+            or an array with the shape (num_points, ndim).
+        nu : array_like, optional, shape (ndim,)
+            Orders of derivatives to evaluate. Each must be non-negative.
```
Defaults to the zeroth derivative?

ev-br replied:
```suggestion
            Orders of derivatives to evaluate. Each must be non-negative. Defaults to the zeroth derivivative.
```

j-bowhay reviewed (commented):
Some minor points, the rest LGTM

ev-br reviewed (commented):
Thanks Jake, merged in your suggestions.

ev-br commented on scipy/interpolate/_ndbspline.py:
```diff
@@ -0,0 +1,205 @@
+        if len(k) != ndim:
+            raise ValueError(f"{len(t) = } != {len(k) = }.")
+
+        self.k = tuple(operator.index(ki) for ki in k)
```
Yes, that.
It parallels https://github.com/scipy/scipy/blob/main/scipy/interpolate/_bsplines.py#L210 : 

```
In [5]: operator.index(3.4)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-f6c97cfd906e> in <module>
----> 1 operator.index(3.4)

TypeError: 'float' object cannot be interpreted as an integer

In [6]: operator.index([1, 2, 3])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-c230828d8837> in <module>
----> 1 operator.index([1, 2, 3])

TypeError: 'list' object cannot be interpreted as an integer
```

Kai-Striega commented:
Merged :) Thanks for your work @ev-br, @AtsushiSakai and @j-bowhay!

kcharlie2 commented:
This looks excellent - I have a couple questions about potential features:

1. Could PChip interpolation be made compatible with this to avoid recursive 1D spline creation? I think this would be a very beneficial capability for users due to the "nice" monotone behavior of pchip interpolation. Unfortunately I don't have the skills to figure out how to implement this.
2. Will there be an ability to modify properties of the NdBSpline object, for example to change extrapolation behavior, as done in the following example using the .extend method? https://docs.scipy.org/doc/scipy/tutorial/interpolate/extrapolation_examples.html#cubicspline-extend-the-boundary-conditions
